libdrgn/program.c

```diff
@@ -868,6 +870,28 @@ static struct drgn_error *get_prpsinfo_pid(struct drgn_program *prog,
 	return NULL;
 }
 
+static struct drgn_error *get_prpsinfo_fname(struct drgn_program *prog,
+					   const char *data, size_t size,
+					   char **ret)
+{
+	bool is_64_bit;
+	struct drgn_error *err = drgn_program_is_64_bit(prog, &is_64_bit);
+	if (err)
+		return err;
+
+	size_t offset = is_64_bit ? 40 : 28;
+	char pr_fname[TASK_COMM_LEN];
+	if (size < offset + sizeof(pr_fname)) {
+		return drgn_error_create(DRGN_ERROR_OTHER,
+					 "NT_PRPSINFO is truncated");
+	}
+	memcpy(pr_fname, data + offset, sizeof(pr_fname));
+	*ret = strndup(pr_fname, sizeof(pr_fname));
```

Comment:
We actually don't need to make a copy here: the data returned by `elf_getdata_rawchunk()` is valid for the lifetime of the `Elf` handle, and `prog->core` is valid for the lifetime of `prog`. So all we need to do here is check that the size is sufficient and that the name is null-terminated, then we can just use the pointer directly.

Btw, let's hard-code 16 for that instead of using `TASK_COMM_LEN` to indicate that the length is defined by the note format. The kernel source code does the same: https://github.com/torvalds/linux/blob/master/include/linux/elfcore.h#L68-L73